service/matching/matchingEngine.go

```diff
@@ -795,19 +815,25 @@ func (e *matchingEngineImpl) GetTaskQueueMetadata(
 	if err != nil {
 		return nil, err
 	}
-	resp := &matchingservice.GetTaskQueueMetadataResponse{}
-	verDatHash := req.GetWantVersioningDataCurhash()
-	// This isn't != nil, because gogoproto will round-trip serialize an empty byte array in a request
-	// into a nil field.
-	if len(verDatHash) > 0 {
-		vDat, err := tqMgr.GetVersioningData(hCtx)
-		if err != nil {
-			return nil, err
+	version := req.GetLastKnownUserDataVersion()
+	if version < 0 {
+		return nil, serviceerror.NewInvalidArgument("last_known_user_data_version must not be negative")
+	}
+	resp := &matchingservice.GetTaskQueueUserDataResponse{}
+	userData, err := tqMgr.GetUserData(hCtx)
+	if err != nil {
+		return nil, err
+	}
+	if userData != nil {
+		resp.TaskQueueHasUserData = true
+		if userData.Version > version {
+			resp.UserData = userData
 		}
-		if !bytes.Equal(HashVersioningData(vDat), verDatHash) {
-			resp.VersioningDataResp = &matchingservice.GetTaskQueueMetadataResponse_VersioningData{
-				VersioningData: vDat,
-			}
+		if userData.Version < version {
+			e.logger.Error("Non-root partition requested task queue user data for version greater than known version")
+			// TODO: When would this happen? Likely due to ownership transfer of the root partition.
```

Comment:
hmm. this shouldn't really be possible except with a very unlikely series of events. e.g. A had the root and loaded user data with version 3. ownership changed, B had the root, some client changed the data to version 4. C (a leaf of that root) refreshed from B. A happened to not notice it lost ownership in all this time. ownership changed again, back to A. now C refreshes from A.

if all the refresh intervals are the same, A should have refreshed in between the two refreshes of C, so that couldn't happen as written. but perhaps it got a timeout that time.

I suppose A could use it as a signal that it has out of date info. but it may be annoying to trigger a reload from this point. I think it would be ok to just return an error and let the periodic refresh fix it.

Reply:
SG